DR2-1182 feat: Use BatchGetItem and add BatchPutItems

MancunianSam opened the pull request:
I've added a new putItems method for writing entries to Dynamo and I've
updated the get items method to use the BatchGetItem.

I've used scanamo DynamoFormat implicits to map case classes to dynamo
values and dynamo values to case classes.

techncl commented on site-docs/src/main/paradox/dynamodb/usage/index.md:
```diff
@@ -9,28 +9,40 @@ val clientWithCustom = new DADynamoDBClient(dynamoDbAsyncClient)
+
+The case class for the primary key should match the primary key of the dynamo table. 
+
+If the table has a primary key of id of type string, the case class would be `PrimaryKey(id: String)`
```
Maybe "If, for example, the table has a primary key called "id", of type String,..."

techncl commented on site-docs/src/main/paradox/dynamodb/usage/index.md:
```diff
@@ -9,28 +9,40 @@ val clientWithCustom = new DADynamoDBClient(dynamoDbAsyncClient)
+The case class for the primary key should match the primary key of the dynamo table. 
+
+If the table has a primary key of id of type string, the case class would be `PrimaryKey(id: String)`
+If there is a composite key of id of type String and index of type number, the case class would be `PrimaryKey(id: String, index: Long)`
```
Maybe "If, for example, there is a composite key of "id", of type String and "index"..."

techncl commented on dynamodb/src/test/scala/uk/gov/nationalarchives/DADynamoDBClientTest.scala:
```diff
@@ -204,4 +241,84 @@ class DADynamoDBClientTest extends AnyFlatSpec with MockitoSugar {
+
+  val putItemsTable: TableFor2[List[MockResponse], FieldName] = Table(
+    ("input", "expectedResponse"),
+    (List(MockSingleAttributeResponse("mockValue")), "{mockAttribute=AttributeValue(S=mockValue)}"),
```
Are the uses of `MockXXResponse`s in this `Table`, here correct? It looks different from the `getItems` one.

MancunianSam replied:
Yeah, I think so. The table is different to the getItems one but it's a different request.

techncl commented on dynamodb/src/test/scala/uk/gov/nationalarchives/DADynamoDBClientTest.scala:
```diff
@@ -6,118 +6,155 @@ import org.mockito.ArgumentMatchers.any
+  forAll(primaryKeysTable) { (keys, expectedResponse) =>
+    "getAttributeValues" should s"pass in $expectedResponse to the BatchGetItemRequest" in {
+      val mockDynamoDbAsyncClient = mock[DynamoDbAsyncClient]
+      val response = Map("mockAttribute" -> AttributeValue.builder().s("mockAttributeValue").build()) asJava
```
You have the infix notation version for the `asJava` here

techncl reviewed (commented):
Looks good; just a few comments. Also, should we add `getItems` tests for these cases:

1. NoPropertyOfType
2. TypeCoercionError
3. MissingProperty

MancunianSam commented:
I've tried to recreate the missing property error but this just gets turned into a `InvalidPropertiesError` and this is checked for in the  `return an error if there are nested field values missing` test.
I've added in a test for `NoPropertyOfType` `return an error if the value is of an unexpected type` 
I can't get it to recreate the `TypeCoercionError` either. I suspect that only happens when you try to do something we're not doing. 

I've also updated it to use the show method on the `DynamoReadError` class which removes the need to have our own case statement on the case classes.

MancunianSam commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +24,96 @@ import scala.language.postfixOps
+      .map(format.write)
+      .map(_.toAttributeValue.m())
+      .map(itemMap => {
+        val pir = PutRequest.builder().item(itemMap).build
```
I think it was originally putItemRequest but I've changed it to `putRequest`

MancunianSam commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +24,96 @@ import scala.language.postfixOps
+  def putItems[T <: Product](tableName: String, items: List[T])(implicit
+      format: DynamoFormat[T]
+  ): F[BatchWriteItemResponse] = {
+    val values: List[WriteRequest] = items
```
`valuesToWrite` makes sense. I've changed it.

MancunianSam commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +24,96 @@ import scala.language.postfixOps
+        case TypeCoercionError(t) => t
+        case MissingProperty      => new Exception("There is an unknown missing property")
+        case InvalidPropertiesError(errors) =>
+          new Exception(s"The following properties are invalid ${errors.map(_._1).toList.mkString}")
```
This is using the `show` method now so the library is taking care of the error messages.

MancunianSam commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +24,96 @@ import scala.language.postfixOps
-      .tableName(dynamoDbRequest.tableName)
-      .key(dynamoDbRequest.primaryKeyAndItsValue asJava)
-      .build()
+  def putItems[T <: Product](tableName: String, items: List[T])(implicit
```
Yeah ok, I've changed it here and in the documentation.

MancunianSam commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +24,96 @@ import scala.language.postfixOps
+    * @return
+    *   A list of case classes of type T which the values populated from Dynamo, wrapped with F[_]
+    */
+  def getItems[T <: Product, K <: Product](keys: List[K], tableName: String)(implicit
```
I've changed it and I've changed the `primaryKeys` variable in the method to `primaryKeysAsAttributeValues`

MancunianSam commented on dynamodb/src/test/scala/uk/gov/nationalarchives/DADynamoDBClientTest.scala:
```diff
@@ -204,4 +241,84 @@ class DADynamoDBClientTest extends AnyFlatSpec with MockitoSugar {
+    override def read(av: DynamoValue): Either[DynamoReadError, MockResponse] = Right(MockSingleAttributeResponse(""))
+
+    override def write(t: MockResponse): DynamoValue = t match {
+      case MockSingleAttributeResponse(mockAttribute) =>
```
Ah good spot, yeah.

MancunianSam commented on dynamodb/src/test/scala/uk/gov/nationalarchives/DADynamoDBClientTest.scala:
```diff
@@ -204,4 +241,84 @@ class DADynamoDBClientTest extends AnyFlatSpec with MockitoSugar {
+      val keySet = value.requestItems().keySet().asScala
+      keySet.size should equal(1)
+      keySet.head should equal("table")
+      val writeValues = value.requestItems().get(keySet.head).asScala
```
`batchGetItemRequestValuesToWrite` is good.

techncl replied:
Sorry, my misspelling has had a knock-on effect; should this just be `batchWriteItemRequestValues`?

MancunianSam commented on dynamodb/src/test/scala/uk/gov/nationalarchives/DADynamoDBClientTest.scala:
```diff
@@ -204,4 +241,84 @@ class DADynamoDBClientTest extends AnyFlatSpec with MockitoSugar {
+
+      putCaptor.getAllValues.size should equal(1)
+      val value = putCaptor.getAllValues.asScala.head
+      val keySet = value.requestItems().keySet().asScala
```
I've gone with `batchGetItemRequestKeys` as there's potentially more than one.

techncl replied:
Sorry, my misspelling has had a knock-on effect; this is supposed to be `Write` instead of `Get`, right?

MancunianSam replied:
Good point, I've changed it.

techncl commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +25,88 @@ import scala.language.postfixOps
-      .key(dynamoDbRequest.primaryKeyAndItsValue asJava)
-      .build()
-
+  def writeItems[T <: Product](tableName: String, items: List[T])(implicit
```
I was thinking, should we add "batch" to be the beginning of it to differentiate it from the non-batch method(s)?

MancunianSam replied:
Same comment as above.

MancunianSam commented on dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:
```diff
@@ -23,32 +25,88 @@ import scala.language.postfixOps
+    * @return
+    *   A list of case classes of type T which the values populated from Dynamo, wrapped with F[_]
+    */
+  def getItems[T <: Product, K <: Product](primaryKeys: List[K], tableName: String)(implicit
```
I think we should leave it. There's a (small) possibility there's some sort of cost/performance penalty for using the batch operations so we may switch this in the future to only use batch when more than one key is provided. We'd want that to be transparent to the calling client though.
